src/MySqlConnector/Serialization/MySqlSession.cs

```diff
@@ -397,6 +370,112 @@ public void FinishQuerying()
+			}
+
+			// add NUL terminator to password
+			var passwordBytes = Encoding.UTF8.GetBytes(cs.Password);
+			Array.Resize(ref passwordBytes, passwordBytes.Length + 1);
+
+			using (rsa)
+			{
+				// XOR the password bytes with the challenge
+				AuthPluginData = Utility.TrimZeroByte(switchRequest.Data);
+				for (var i = 0; i < passwordBytes.Length; i++)
+					passwordBytes[i] ^= AuthPluginData[i % AuthPluginData.Length];
+
+				// encrypt with RSA public key
+				var encryptedPassword = rsa.Encrypt(passwordBytes, RSAEncryptionPadding.Pkcs1);
+				var payload = new PayloadData(new ArraySegment<byte>(encryptedPassword));
+				await SendReplyAsync(payload, ioBehavior, cancellationToken).ConfigureAwait(false);
+				return await ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+			}
+		}
+#endif
+
+		private async Task<string> GetRsaPublicKeyForSha256PasswordAsync(
+			string switchRequestName,
+			ConnectionSettings cs,
+			IOBehavior ioBehavior,
+			CancellationToken cancellationToken)
+		{
+			if (!string.IsNullOrEmpty(cs.ServerRsaPublicKeyFile))
+			{
+				try
+				{
+					return File.ReadAllText(cs.ServerRsaPublicKeyFile);
+				}
+				catch (IOException ex)
+				{
+					throw new MySqlException(
+						"Couldn't load server's RSA public key from '{0}'".FormatInvariant(cs.ServerRsaPublicKeyFile), ex);
+				}
+			}
+
+			if (cs.AllowPublicKeyRetrieval)
+			{
+				// request the RSA public key
+				await SendReplyAsync(new PayloadData(new ArraySegment<byte>(new byte[] { 0x01 }, 0, 1)), ioBehavior,
+					cancellationToken).ConfigureAwait(false);
+				var payload = await ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+				var publicKeyPayload = AuthenticationMoreDataPayload.Create(payload);
+				return Encoding.ASCII.GetString(publicKeyPayload.Data);
+			}
+
+			throw new MySqlException(
+				"Authentication method '{0}' failed. Either use a secure connection, specify the server's RSA public key with ServerRSAPublicKeyFile, or set AllowPublicKeyRetrieval=True."
+					.FormatInvariant(switchRequestName));
+		}
+
+		private async Task<string> GetRsaPublicKeyForCachingSha2PasswordAsync(
+			string switchRequestName,
+			ConnectionSettings cs,
+			IOBehavior ioBehavior,
+			CancellationToken cancellationToken)
+		{
+			if (cs.AllowPublicKeyRetrieval)
+			{
+				// request the RSA public key
+				await SendReplyAsync(new PayloadData(new ArraySegment<byte>(new byte[] { 0x02 }, 0, 1)), ioBehavior,
+					cancellationToken).ConfigureAwait(false);
+				var payload = await ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+				var publicKeyPayload = AuthenticationMoreDataPayload.Create(payload);
+				return Encoding.ASCII.GetString(publicKeyPayload.Data);
+			}
+
+			throw new MySqlException(
+				"Authentication method '{0}' failed. Either use a secure connection, specify the server's RSA public key with ServerRSAPublicKeyFile, or set AllowPublicKeyRetrieval=True."
```

Comment:
Does using `ServerRSAPublicKeyFile` work for `caching_sha256_password`? This method doesn't appear to load from the file specified by that connection string setting, but the exception message still mentions it.

Reply:
(Does MySQL Server use the same RSA public key for `sha256_password` and `caching_sha2_password` or can it support a different public/private keypair for each authentication method?)

Reply:
It can support different key-pair for each method:
- `caching_sha2_password_private_key_path`, `caching_sha2_password_public_key_path` for `caching_sha2_password`;
- `sha256_password_private_key_path`, `sha256_password_public_key_path` for `sha256_password`.

Reply:
>Does using ServerRSAPublicKeyFile work for caching_sha2_password? This method doesn't appear to load from the file specified by that connection string setting, but the exception message still mentions it.

I removed support of ServerRSAPublicKeyFile, but forgot to fix exception text. Should I add another specific option for the file or fix exception message?

Reply:
Please remove it from the exception message for now.

Long-term, it seems like another connection string option should be provided, but it could be good to wait to see what other connector libraries do (for consistency). For example, Connector/J doesn't seem to have a different option in [its configuration properties](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-configuration-properties.html) yet.

Reply:
Actually (after thinking about it more), from the client side there doesn't seem to be any reason we couldn't use `ServerRSAPublicKeyFile` for both `sha256_password` and `caching_sha2_password`. Presumably when the user creates the connection string (with the user ID and password) they know which auth plugin the server is using, and which RSA private key it's set up with. For any individual connection, there's no need to specify both the `sha256_password` public key *and* the `caching_sha2_password` public key. Thus, the same connection string option can be used for both. More specifically, `ServerRSAPublicKeyFile` can continue to be used in this new code path to specify the `caching_sha2_password` public key file.

Reply:
I think in `caching_sha2_password` scenario there are 2 cases:
1. Server has RSA keys, and always can provide them (does not require OpenSSL);
2. Server has not any RSA keys, so there is no point in specifying public key at client-side.

Reply:
The point of specifying them client-side is to prevent a MITM attack; see #286.

Reply:
Thank you. I will fix it by restoring `ServerRSAPublicKeyFile` support.